p2p/service.py

```diff
@@ -117,6 +124,24 @@ def run_child_service(self, child_service: 'BaseService') -> 'asyncio.Future[Any
     def is_running(self) -> bool:
         return self._run_lock.locked()
 
+    async def threadsafe_cancel(self, poll_period: float=0.1) -> None:
+        """
+        Cancel service in another thread. Block until service is cleaned up.
+
+        :param poll_period: how many seconds to wait in between each check for service cleanup
+        """
+        asyncio.run_coroutine_threadsafe(self.cancel(), loop=self.loop)
+
+        start_time = time.monotonic()
+        while time.monotonic() - start_time < self._wait_until_finished_timeout:
+            if not self.cleaned_up.is_set():
```

Comment:
I think you can drop the whole loop here and replace it with:

```
asyncio.wait_for(self.cleaned_up.wait(), timeout=self._wait_until_finished_timeout)
```

This should allow `asyncio` to manage the timeout.

Reply:
Hm, not that the current implementation necessarily solves this, but should we be worried that [`asyncio.Event` is marked as "not thread safe"](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event)?

I tried it out, and seems to work okay. As long as it's safe, it's clearly better (less code).